### PR TITLE
Add timers and fix set usage

### PR DESCRIPTION
**Pull request by EmilyBourne**
In order to increase the speed of the Pyccel command (see #1030 , #1608 and discussions on discord with @dominikbell ) it is important to understand which sections of the code represent the bottlenecks. This PR adds some timers around the main parts of the code and adds a flag which allows the results to be printed.

This PR identifies the wrapper creation as a bottleneck although it is unclear why this is so much more expensive than the semantic stage. A quick investigation showed that the misuse of sets in the `naming` module is part of the cause of this. This PR fixes this and therefore gains a significant speed up (~x2) in the wrapper creation.

On my computer the example provided by @dominikbell takes 42s before this PR (17s for C) and 36s after (14s for C).

**Commit Summary**
- Add a `time_execution` flag to allow users to time the execution of critical parts
- Use `a in set` syntax to benefit from set optimisations
- Only use `.lower()` when testing collisions in Fortran code
- Ensure temporaries are not renamed on insertion to scope of module initialisation function.

### CODE REVIEW

**Comment by EmilyBourne** [comment]
/bot run docs coverage pylint pyccel_lint spelling

File: pyccel/codegen/pipeline.py [hunk 1]
```diff
@@ -212,14 +218,16 @@ def handle_error(stage):
         handle_error('parsing (syntax)')
         raise PyccelSyntaxError('Syntax step failed')
 
+    end_syntax = time.time()
+
     if syntax_only:
         pyccel_stage.pyccel_finished()
         return
```
Comment (yguclu) [inline pyccel/codegen/pipeline.py:225]: Don't we have a problem here? We are not printing any timings unless we get to the end of this function.

As a workaround I suggest that we update the `timings` dictionary on the fly (the overhead of doing so should be negligible compared to everything else that goes on in Pyccel) and then pass it to a `print_timings` function which prints whatever is available and the total elapsed time.

File: pyccel/codegen/pipeline.py [hunk 2]
```diff
@@ -234,13 +242,16 @@ def handle_error(stage):
         handle_error('annotation (semantic)')
         raise PyccelSemanticError('Semantic step failed')
 
+    end_semantic = time.time()
+
     if semantic_only:
         pyccel_stage.pyccel_finished()
         return
```
Comment (yguclu) [inline pyccel/codegen/pipeline.py:249]: Same comment as above. Other `return` statements are on lines 285, 332, and 343.

**Review by yguclu — CHANGES_REQUESTED** [review]

File: pyccel/codegen/python_wrapper.py [hunk 3]
```diff
@@ -198,9 +216,18 @@ def create_shared_library(codegen,
...
     # Return absolute path of shared library
-    return sharedlib_filepath
+    timings = {'Dependency compilation' : compile_libs_time,
+               'Wrapper creation' : wrapper_creation_time,
+               'Wrapper printing': print_cwrapper_time,
+               'Wrapper compilation': compile_wrapper_time}
+    if language == 'fortran':
+        timings['Bind C wrapping'] = bind_c_wrapping_time
+        timings['Bind C printing'] = bind_c_printing_time
+        timings['Bind C compiling'] = bind_c_compiling_time
```
Comment (yguclu) [inline pyccel/codegen/python_wrapper.py:232]: The new code should be added above the comment line `# Return absolute path of shared library`.

By the way, `timings` should be added to the `Returns` section of the function docstring.

File: pyccel/naming/fortrannameclashchecker.py [hunk 4]
```diff
@@ -48,10 +48,27 @@ class FortranNameClashChecker(LanguageNameClashChecker):
...
+
+        Returns
+        -------
+        bool
+            True if the name clashes with an existing name. False otherwise.
         """
         name = name.lower()
-        return any(name == k for k in self.keywords) or \
+        return name in self.keywords or \
                any(name == s.lower() for s in symbols)
```
Comment (yguclu) [inline pyccel/naming/fortrannameclashchecker.py:72]: The names in `symbols` are the original Python names, not converted to lowercase?

**Review by yguclu — CHANGES_REQUESTED** [review]

File: pyccel/naming/fortrannameclashchecker.py (hunk 4 above)
Comment (EmilyBourne) [reply pyccel/naming/fortrannameclashchecker.py:72]: After this PR, yes. Before they were converted to lowercase (at a non-negligible cost) but this was only useful for Fortran.

File: pyccel/codegen/python_wrapper.py (hunk 3 above)
Comment (EmilyBourne) [reply pyccel/codegen/python_wrapper.py:234]: Fixed f679fca1

Comment (EmilyBourne) [reply pyccel/codegen/python_wrapper.py:234]: I'm surprised it didn't complain

File: pyccel/codegen/pipeline.py (hunk 1 above)
Comment (EmilyBourne) [reply pyccel/codegen/pipeline.py:231]: No but I think documenting it will be :laughing:

Comment (EmilyBourne) [reply pyccel/codegen/pipeline.py:231]: Can I put it as a function in a function?

Comment (yguclu) [reply pyccel/codegen/pipeline.py:231]: Sure

File: pyccel/codegen/pipeline.py [hunk 5]
```diff
@@ -317,6 +338,8 @@
         # Change working directory back to starting point
         os.chdir(base_dirpath)
         pyccel_stage.pyccel_finished()
+        if show_timings:
+            print_timers(start, timers)
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/pipeline.py:342]: /bot accept
The function is tested.

File: pyccel/codegen/pipeline.py [hunk 6]
```diff
@@ -270,6 +289,8 @@
...
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/pipeline.py:293]: /bot accept
The function is tested.

File: pyccel/codegen/pipeline.py [hunk 7]
```diff
@@ -234,13 +246,18 @@
         handle_error('annotation (semantic)')
         raise PyccelSemanticError('Semantic step failed')
 
+    timers["Semantic Stage"] = time.time() - start_semantic
+
     if semantic_only:
         pyccel_stage.pyccel_finished()
+        if show_timings:
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/pipeline.py:253]: /bot accept
The function is tested.

File: pyccel/codegen/pipeline.py [hunk 8]
```diff
@@ -212,14 +220,18 @@ def handle_error(stage):
         handle_error('parsing (syntax)')
         raise PyccelSyntaxError('Syntax step failed')
 
+    timers["Syntactic Stage"] = time.time() - start_syntax
+
     if syntax_only:
         pyccel_stage.pyccel_finished()
+        if show_timings:
+            print_timers(start, timers)
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/pipeline.py:228]: /bot accept
The function is tested.

**Review by yguclu — APPROVED** [review]

**Pull request merged**